### PR TITLE
Uploadscript

### PR DESCRIPTION
**Pull request by Ma-Shell**
Improvements to the upload-script:
- Parallellized with a configurable number of worker-threads.
- MIME-type-checking
- Ability to select a local directory for flat or recursive traversal

### CODE REVIEW

**Comment by cynexit** [comment]
Please use either tabs or four spaces everywhere. :)

File: push_to_holmes.go [hunk 1]
```diff
@@ -50,36 +73,100 @@ func main() {
 	flag.StringVar(&fPath, "file", "", "List of samples (MD5, SHAX, CRITs ID) to copy from CRITs to Totem")
```
Comment (cynexit) [inline push_to_holmes.go]: Please clear up: Not only from CRITs but from various sources.

File: push_to_holmes.go [hunk 2]
```diff
@@ -14,8 +14,13 @@ import (
 	"net/http"
 	"os"
 	"time"
+	"path/filepath"
 
 	"gopkg.in/mgo.v2/bson"
+	"sync"
+
+	"github.com/rakyll/magicmime"
```
Comment (cynexit) [inline push_to_holmes.go]: Maybe simply use http.DetectContentType, discuss. This would remove the libmagic dependency.

File: push_to_holmes.go [hunk 3]
```diff
@@ -50,36 +73,100 @@ func main() {
...
+        return
+    }
+
+	wg.Wait()
+}
 
-	// line by line
-	for scanner.Scan() {
-		copySample(scanner.Text())
+func walkFn(path string, fi os.FileInfo, err error) (e error) {
```
Comment (cynexit) [inline push_to_holmes.go]: (e error) -> error

File: push_to_holmes.go [hunk 4]
```diff
@@ -50,36 +73,100 @@ func main() {
...
+			}
+		}
+	}
+	mimetype, err := magicmime.TypeByFile(path)
+	if err != nil {
+		log.Println("mimetype error (skipping " + path + "):", err)
+		return nil
+	}
+	if strings.Contains(mimetype, "exec") {
+		print("Adding " + path + " (" + mimetype + ")\n")
```
Comment (cynexit) [inline push_to_holmes.go]: Please use log.Println

File: push_to_holmes.go [hunk 5]
```diff
@@ -50,36 +73,100 @@ func main() {
...
+		wg.Add(1)
+		c <- path
+		return nil
+	} else {
+		print("Skipping " + path + " (" + mimetype + ")\n")
```
Comment (cynexit) [inline push_to_holmes.go]: Please use log.Println

File: push_to_holmes.go (hunk 2 above)
Comment (Ma-Shell) [inline push_to_holmes.go]: After fiddling around with http.DetectContentType, I found that http.DetectContentType performs vastly worse than magicmime, especially when it comes to executable files (which should be the focus here). For all of my tests magicmime classified them correctly as "application/x-dosexec" or "application/x-executable", while http.DetectContentType just said "application/octet-stream", which is the fallback, if nothing else is recognized. For this reason I would bite the dependency-bullet and settle for magicmime.

Comment (webstergd) [inline push_to_holmes.go]: He verified this with me. http.DetectContentType is shockingly bad.

**Comment by webstergd** [comment]
Danke

**Pull request merged**